RecoTracker/LSTCore/src/alpaka/Event.dev.cc

```diff
@@ -0,0 +1,1824 @@
+  }
+  if (pixelTripletsInCPU != nullptr) {
+    delete pixelTripletsInCPU;
+    pixelTripletsInCPU = nullptr;
+  }
+  if (pixelQuintupletsInCPU != nullptr) {
+    delete pixelQuintupletsInCPU;
+    pixelQuintupletsInCPU = nullptr;
+  }
+  if (trackCandidatesInCPU != nullptr) {
+    delete trackCandidatesInCPU;
+    trackCandidatesInCPU = nullptr;
+  }
+  if (modulesInCPU != nullptr) {
+    delete modulesInCPU;
+    modulesInCPU = nullptr;
+  }
+}
+
+void SDL::Event<SDL::Acc>::addHitToEvent(std::vector<float> x,
+                                         std::vector<float> y,
+                                         std::vector<float> z,
+                                         std::vector<unsigned int> detId,
+                                         std::vector<unsigned int> idxInNtuple) {
+  // Use the actual number of hits instead of a max.
+  unsigned int nHits = x.size();
+
+  // Initialize space on device/host for next event.
+  if (hitsInGPU == nullptr) {
+    hitsInGPU = new SDL::hits();
+    hitsBuffers = new SDL::hitsBuffer<Dev>(nModules_, nHits, devAcc, queue);
+    hitsInGPU->setData(*hitsBuffers);
+  }
+
+  if (rangesInGPU == nullptr) {
+    rangesInGPU = new SDL::objectRanges();
+    rangesBuffers = new SDL::objectRangesBuffer<Dev>(nModules_, nLowerModules_, devAcc, queue);
+    rangesInGPU->setData(*rangesBuffers);
+  }
+
+  // Need a view here before transferring to the device.
+  auto nHits_view = alpaka::createView(devHost, &nHits, (Idx)1u);
+
+  // Copy the host arrays to the GPU.
+  alpaka::memcpy(queue, hitsBuffers->xs_buf, x, nHits);
+  alpaka::memcpy(queue, hitsBuffers->ys_buf, y, nHits);
+  alpaka::memcpy(queue, hitsBuffers->zs_buf, z, nHits);
+  alpaka::memcpy(queue, hitsBuffers->detid_buf, detId, nHits);
+  alpaka::memcpy(queue, hitsBuffers->idxs_buf, idxInNtuple, nHits);
+  alpaka::memcpy(queue, hitsBuffers->nHits_buf, nHits_view);
+  alpaka::wait(queue);
+
+  Vec const threadsPerBlock1 = createVec(1, 1, 256);
+  Vec const blocksPerGrid1 = createVec(1, 1, MAX_BLOCKS);
+  WorkDiv const hit_loop_workdiv = createWorkDiv(blocksPerGrid1, threadsPerBlock1, elementsPerThread);
+
+  hitLoopKernel hit_loop_kernel;
+  auto const hit_loop_task(alpaka::createTaskKernel<Acc>(hit_loop_workdiv,
+                                                         hit_loop_kernel,
+                                                         Endcap,
+                                                         TwoS,
+                                                         nModules_,
+                                                         nEndCapMap_,
+                                                         alpaka::getPtrNative(endcapGeometryBuffers_->geoMapDetId_buf),
+                                                         alpaka::getPtrNative(endcapGeometryBuffers_->geoMapPhi_buf),
+                                                         *modulesBuffers_->data(),
+                                                         *hitsInGPU,
```

Comment:
Following my comment on `hits` and `hitsBuffer` relation, (in the present design) couldn't this be just `*hitsBuffers` (relying on slicing of the base class part)?

Anyway, I'd find the "owner" and "non-owning view" separation much more clear. In that approach the `hitsInGPU` member would not be needed, and this line would be along `hitsBuffers->view()`.

Reply:
> Following my comment on `hits` and `hitsBuffer` relation, (in the present design) couldn't this be just `*hitsBuffers` (relying on slicing of the base class part)?

Ah, I see with `hitsBuffer` this would not work out of the box because `hitsBuffer` constructor does not call `hits::setData()` (whereas e.g. `modulesBuffer` constructor calls the `setData()` of its base class).